From 6c80836fac4e8da375300756b35ab641dc52a165 Mon Sep 17 00:00:00 2001
From: Pierre-Hugues Husson <phh@phh.me>
Date: Sun, 19 Dec 2021 17:08:22 -0500
Subject: [PATCH 26/35] Add UDFPS support for various devices

Successfully tested on Galaxy A51, Xiaomi Mi 9, and OnePlus Nord 2

- Detect UDFPS properties (udfps support and x/y/r) based on
  OEM-specific methods
- Hardcode UDFPS properties for some devices
- Call OEM-specific functions on touchDown/touchUp

Change-Id: I9448f06a46c178652f5b216717d914b8db55c993
---
 .../systemui/biometrics/UdfpsSurfaceView.java |   2 +-
 .../systemui/biometrics/UdfpsView.java        |  60 +++-
 services/core/Android.bp                      |   4 +
 .../server/biometrics/AuthService.java        | 262 +++++++++++++++++-
 .../sensors/fingerprint/UdfpsHelper.java      | 115 +++++++-
 5 files changed, 423 insertions(+), 20 deletions(-)

diff --git a/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsSurfaceView.java b/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsSurfaceView.java
index 77fad35d32d4..51409b9c533a 100644
--- a/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsSurfaceView.java
+++ b/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsSurfaceView.java
@@ -69,7 +69,7 @@ public class UdfpsSurfaceView extends SurfaceView implements SurfaceHolder.Callb
 
         mSensorPaint = new Paint(0 /* flags */);
         mSensorPaint.setAntiAlias(true);
-        mSensorPaint.setARGB(255, 255, 255, 255);
+        mSensorPaint.setARGB(255, 0, 255, 0);
         mSensorPaint.setStyle(Paint.Style.FILL);
     }
 
diff --git a/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsView.java b/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsView.java
index 6d31ef0e7701..dd5da24a87bf 100644
--- a/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsView.java
+++ b/packages/SystemUI/src/com/android/systemui/biometrics/UdfpsView.java
@@ -42,6 +42,12 @@ import com.android.systemui.R;
 import com.android.systemui.biometrics.UdfpsHbmTypes.HbmType;
 import com.android.systemui.doze.DozeReceiver;
 
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.PrintWriter;
+
+
 /**
  * A view containing 1) A SurfaceView for HBM, and 2) A normal drawable view for all other
  * animations.
@@ -51,7 +57,7 @@ public class UdfpsView extends FrameLayout implements DozeReceiver, UdfpsIllumin
 
     private static final String SETTING_HBM_TYPE =
             "com.android.systemui.biometrics.UdfpsSurfaceView.hbmType";
-    private static final @HbmType int DEFAULT_HBM_TYPE = UdfpsHbmTypes.LOCAL_HBM;
+    private static final @HbmType int DEFAULT_HBM_TYPE = UdfpsHbmTypes.GLOBAL_HBM;
 
     private static final int DEBUG_TEXT_SIZE_PX = 32;
 
@@ -71,6 +77,49 @@ public class UdfpsView extends FrameLayout implements DozeReceiver, UdfpsIllumin
     @Nullable private String mDebugMessage;
     private boolean mIlluminationRequested;
 
+    public static void samsungCmd(String cmd) {
+	    try {
+		    writeFile("/sys/devices/virtual/sec/tsp/cmd", cmd);
+
+		    String status = readFile("/sys/devices/virtual/sec/tsp/cmd_status");
+		    String ret = readFile("/sys/devices/virtual/sec/tsp/cmd_result");
+
+		    android.util.Log.d("PHH", "Sending command " + cmd + " returned " + ret + ":" + status);
+	    } catch(Exception e) {
+		    android.util.Log.d("PHH", "Failed sending command " + cmd, e);
+	    }
+    }
+    private static void writeFile(String path, String value) {
+	    try {
+		    PrintWriter writer = new PrintWriter(path, "UTF-8");
+		    writer.println(value);
+		    writer.close();
+	    } catch(Exception e) {
+		    android.util.Log.d("PHH", "Failed writing to " + path + ": " + value);
+	    }
+    }
+
+    private static void writeFile(File file, String value) {
+	    try {
+		    PrintWriter writer = new PrintWriter(file, "UTF-8");
+		    writer.println(value);
+		    writer.close();
+	    } catch(Exception e) {
+		    android.util.Log.d("PHH", "Failed writing to " + file + ": " + value);
+	    }
+    }
+    private static String readFile(String path) {
+	    try {
+		    File f = new File(path);
+
+		    BufferedReader b = new BufferedReader(new FileReader(f));
+		    return b.readLine();
+	    } catch(Exception e) {
+		    return null;
+	    }
+    }
+
+
     public UdfpsView(Context context, AttributeSet attrs) {
         super(context, attrs);
 
@@ -255,6 +255,11 @@ public class UdfpsView extends FrameLayout implements DozeReceiver, UdfpsIllumin
                 }
             });
         }
+	else {
+		if (mGhbmView != null) {
+			mGhbmView.drawIlluminationDot(mSensorRect);
+		}
+	}
     }
 
     @Override
diff --git a/services/core/Android.bp b/services/core/Android.bp
index e5716eee467e..9c359cea6244 100644
--- a/services/core/Android.bp
+++ b/services/core/Android.bp
@@ -155,6 +155,10 @@ java_library_static {
         "android.hardware.soundtrigger-V2.3-java",
         "android.hardware.power.stats-V1-java",
         "android.hidl.manager-V1.2-java",
+        "vendor.goodix.hardware.biometrics.fingerprint-V2.1-java",
+        "vendor.samsung.hardware.biometrics.fingerprint-V3.0-java",
+        "vendor.oplus.hardware.biometrics.fingerprint-V2.1-java",
+        "vendor.oppo.hardware.biometrics.fingerprint-V2.1-java",
         "capture_state_listener-aidl-java",
         "icu4j_calendar_astronomer",
         "netd-client",
diff --git a/services/core/java/com/android/server/biometrics/AuthService.java b/services/core/java/com/android/server/biometrics/AuthService.java
index 9c8ccd946b7f..62ea6dd8633b 100644
--- a/services/core/java/com/android/server/biometrics/AuthService.java
+++ b/services/core/java/com/android/server/biometrics/AuthService.java
@@ -73,6 +73,21 @@ import com.android.server.SystemService;
 import java.util.ArrayList;
 import java.util.List;
 
+import android.hardware.display.DisplayManager;
+
+import android.graphics.Point;
+import android.util.DisplayMetrics;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.PrintWriter;
+
+import android.os.FileObserver;
+import android.os.Build;
+
+import vendor.samsung.hardware.biometrics.fingerprint.V3_0.ISehBiometricsFingerprint;
+import vendor.goodix.hardware.biometrics.fingerprint.V2_1.IGoodixFingerprintDaemon;
+
 /**
  * System service that provides an interface for authenticating with biometrics and
  * PIN/pattern/password to BiometricPrompt and lock screen.
@@ -89,6 +104,9 @@ public class AuthService extends SystemService {
     @VisibleForTesting
     final IAuthService.Stub mImpl;
 
+    private FileObserver fodFileObserver = null;
+    private ISehBiometricsFingerprint mSamsungFingerprint = null;
+
     /**
      * Class for injecting dependencies into AuthService.
      * TODO(b/141025588): Replace with a dependency injection framework (e.g. Guice, Dagger).
@@ -637,6 +655,68 @@ public class AuthService extends SystemService {
         registerAuthenticators(hidlConfigs);
 
         mInjector.publishBinderService(this, mImpl);
+        try {
+            mSamsungFingerprint = ISehBiometricsFingerprint.getService();
+            android.util.Log.e("PHH", "Got samsung fingerprint HAL");
+        } catch(Exception e) {
+            android.util.Log.e("PHH", "Failed getting Samsung fingerprint HAL", e);
+        }
+        if(samsungHasCmd("fod_enable") && mSamsungFingerprint != null) {
+            samsungCmd("fod_enable,1,1,0");
+            String actualMaskBrightnessPath = "/sys/class/lcd/panel/actual_mask_brightness";
+            android.util.Log.e("PHH-Enroll", "Reading actual brightness file gives " + readFile(actualMaskBrightnessPath));
+            fodFileObserver = new FileObserver(actualMaskBrightnessPath, FileObserver.MODIFY) {
+                @Override
+                public void onEvent(int event, String path) {
+                    String actualMask = readFile(actualMaskBrightnessPath);
+                    try {
+                        mSamsungFingerprint = ISehBiometricsFingerprint.getService();
+                    } catch(Exception e) {}
+                    Slog.d("PHH-Enroll", "New actual mask brightness is " + actualMask);
+                    try {
+                        int eventReq = 0;
+                        if("0".equals(actualMask)) {
+                            eventReq = 1; //released
+                        } else {
+                            eventReq = 2; //pressed
+                        }
+                        if(mSamsungFingerprint != null) {
+                            mSamsungFingerprint.sehRequest(22 /* SEM_FINGER_STATE */, eventReq, new java.util.ArrayList<Byte>(),
+                                    (int retval, java.util.ArrayList<Byte> out) -> {} );
+                        }
+                    } catch(Exception e) {
+                        Slog.d("PHH-Enroll", "Failed setting samsung event for mask observer", e);
+                    }
+                }
+            };
+            fodFileObserver.startWatching();
+        }
+        String asusSpotOnAchieved = "/sys/class/drm/spot_on_achieved";
+        if( (new File(asusSpotOnAchieved)).exists()) {
+            fodFileObserver = new FileObserver(asusSpotOnAchieved, FileObserver.MODIFY) {
+                boolean wasOn = false;
+                @Override
+                public void onEvent(int event, String path) {
+                    String spotOn = readFile(asusSpotOnAchieved);
+                    if("1".equals(spotOn)) {
+                        if(!wasOn) {
+                            try {
+                                IGoodixFingerprintDaemon goodixDaemon = IGoodixFingerprintDaemon.getService();
+                                goodixDaemon.sendCommand(200002, new java.util.ArrayList<Byte>(), (returnCode, resultData) -> {
+                                    Slog.e(TAG, "Goodix send command returned code "+ returnCode);
+                                });
+                            } catch(Throwable t) {
+                                Slog.d("PHH-Enroll", "Failed sending goodix command", t);
+                            }
+                        }
+                        wasOn = true;
+                    } else {
+                        wasOn = false;
+                    }
+                }
+            };
+            fodFileObserver.startWatching();
+        }
     }
 
     /**
@@ -737,23 +737,201 @@ public class AuthService extends SystemService {
                 ? modality : (modality & ~BiometricAuthenticator.TYPE_CREDENTIAL);
     }
 
+        private static String readFile(String path) {
+        try {
+            File f = new File(path);
+
+            BufferedReader b = new BufferedReader(new FileReader(f));
+            return b.readLine();
+        } catch(Exception e) {
+            return null;
+        }
+    }
+
+        private static boolean samsungHasCmd(String cmd) {
+        try {
+            File f = new File("/sys/devices/virtual/sec/tsp/cmd_list");
+            if(!f.exists()) return false;
+
+	    android.util.Log.d("PHH", "Managed to grab cmd list, checking...");
+            BufferedReader b = new BufferedReader(new FileReader(f));
+            String line = null;
+            while( (line = b.readLine()) != null) {
+                if(line.equals(cmd)) return true;
+            }
+	    android.util.Log.d("PHH", "... nope");
+            return false;
+        } catch(Exception e) {
+	    android.util.Log.d("PHH", "Failed reading cmd_list", e);
+            return false;
+        }
+    }
+
+        public static void samsungCmd(String cmd) {
+        try {
+	    writeFile("/sys/devices/virtual/sec/tsp/cmd", cmd);
+
+            String status = readFile("/sys/devices/virtual/sec/tsp/cmd_status");
+            String ret = readFile("/sys/devices/virtual/sec/tsp/cmd_result");
+
+            android.util.Log.d("PHH", "Sending command " + cmd + " returned " + ret + ":" + status);
+        } catch(Exception e) {
+            android.util.Log.d("PHH", "Failed sending command " + cmd, e);
+        }
+    }
+
+    private static void writeFile(String path, String value) {
+        try {
+            PrintWriter writer = new PrintWriter(path, "UTF-8");
+            writer.println(value);
+            writer.close();
+        } catch(Exception e) {
+            android.util.Log.d("PHH", "Failed writing to " + path + ": " + value);
+        }
+    }
+
+    private static void writeFile(File file, String value) {
+        try {
+            PrintWriter writer = new PrintWriter(file, "UTF-8");
+            writer.println(value);
+            writer.close();
+        } catch(Exception e) {
+            android.util.Log.d("PHH", "Failed writing to " + file + ": " + value);
+        }
+    }
 
     private FingerprintSensorPropertiesInternal getHidlFingerprintSensorProps(int sensorId,
             @BiometricManager.Authenticators.Types int strength) {
         // The existence of config_udfps_sensor_props indicates that the sensor is UDFPS.
-        final int[] udfpsProps = getContext().getResources().getIntArray(
+        int[] udfpsProps = getContext().getResources().getIntArray(
                 com.android.internal.R.array.config_udfps_sensor_props);
 
         // Non-empty workaroundLocations indicates that the sensor is SFPS.
         final List<SensorLocationInternal> workaroundLocations =
                 getWorkaroundSensorProps(getContext());
 
-        final boolean isUdfps = !ArrayUtils.isEmpty(udfpsProps);
+        boolean isUdfps = !ArrayUtils.isEmpty(udfpsProps);
 
         // config_is_powerbutton_fps indicates whether device has a power button fingerprint sensor.
         final boolean isPowerbuttonFps = getContext().getResources().getBoolean(
                 R.bool.config_is_powerbutton_fps);
 
+        DisplayManager mDM = (DisplayManager) getContext().getSystemService(Context.DISPLAY_SERVICE);
+        Point displayRealSize = new Point();
+        DisplayMetrics displayMetrics = new DisplayMetrics();
+        mDM.getDisplay(0).getRealSize(displayRealSize);
+        mDM.getDisplay(0).getMetrics(displayMetrics);
+
+        String[] xiaomiLocation = android.os.SystemProperties.get("persist.vendor.sys.fp.fod.location.X_Y", "").split(",");
+        if(xiaomiLocation.length != 2)
+            xiaomiLocation = android.os.SystemProperties.get("persist.sys.fp.fod.location.X_Y", "").split(",");
+        String[] xiaomiSize = android.os.SystemProperties.get("persist.vendor.sys.fp.fod.size.width_height", "").split(",");
+        if(xiaomiSize.length != 2)
+            xiaomiSize = android.os.SystemProperties.get("persist.sys.fp.fod.size.width_height", "").split(",");
+        if(xiaomiSize.length == 2 && xiaomiLocation.length == 2) {
+            udfpsProps = new int[3];
+            udfpsProps[0] = (int)displayRealSize.x/2;
+            udfpsProps[1] = Integer.parseInt(xiaomiLocation[1]);
+            udfpsProps[2] = Integer.parseInt(xiaomiSize[0])/2;
+            udfpsProps[1] += udfpsProps[2];
+            isUdfps = true;
+        }
+
+        if(readFile("/sys/class/fingerprint/fingerprint/position") != null) {
+            android.util.Log.d("PHH", "Samsung fingerprint");
+            String[] fodPositionArray = readFile("/sys/class/fingerprint/fingerprint/position").split(",");
+            float bottomMM = Float.parseFloat(fodPositionArray[0]);
+            float areaSizeMM = Float.parseFloat(fodPositionArray[5]);
+            float bottomInch = bottomMM * 0.0393700787f;
+            float areaSizeInch = areaSizeMM * 0.0393700787f;
+            int bottomPx = (int)(bottomInch * displayMetrics.ydpi);
+            int areaSizePx = (int)(areaSizeInch * displayMetrics.ydpi);
+            float mW = areaSizePx/2;
+            float mH = areaSizePx/2;
+            float mX = displayRealSize.x/2;
+            float mY = displayRealSize.y - bottomPx - mH;
+
+            samsungCmd(String.format("fod_rect,%d,%d,%d,%d", (int)(mX - mW/2), (int)(mY - mW/2), (int)(mX + mW/2), (int)(mY + mW/2)));
+
+            udfpsProps = new int[3];
+            udfpsProps[0] = (int)mX;
+            udfpsProps[1] = (int)mY;
+            udfpsProps[2] = (int)mW;
+            isUdfps = true;
+
+            try {
+                mSamsungFingerprint = ISehBiometricsFingerprint.getService();
+                Slog.d("PHH-Enroll", "Samsung ask for sensor status");
+                mSamsungFingerprint.sehRequest(6, 0, new java.util.ArrayList(), (int retval, java.util.ArrayList<Byte> out) -> {
+                    Slog.d("PHH-Enroll", "Result is " + retval);
+                    for(int i=0; i<out.size(); i++) {
+                        Slog.d("PHH-Enroll", "\t" + i + ":" + out.get(i));
+                    }
+                } );
+                Slog.d("PHH-Enroll", "Samsung ask for sensor brightness value");
+                mSamsungFingerprint.sehRequest(32, 0, new java.util.ArrayList(), (int retval, java.util.ArrayList<Byte> out) -> {
+                    Slog.d("PHH-Enroll", "Result is " + retval);
+                    for(int i=0; i<out.size(); i++) {
+                        Slog.d("PHH-Enroll", "\t" + i + ":" + out.get(i));
+                    }
+                } );
+
+            } catch(Exception e) {
+                Slog.d("PHH-Enroll", "Failed setting samsung3.0 fingerprint recognition", e);
+            }
+        }
+
+        int oppoSize = android.os.SystemProperties.getInt("persist.vendor.fingerprint.optical.iconsize", 0);
+        int oppoLocation = android.os.SystemProperties.getInt("persist.vendor.fingerprint.optical.iconlocation", 0);
+        if(oppoLocation > 0 && oppoSize > 0) {
+            int mW = oppoSize/2;
+            int mH = oppoSize/2;
+
+            Slog.d("PHH-Enroll", "Got Oppo icon location " + oppoLocation);
+            Slog.d("PHH-Enroll", "\tscreen size " + displayRealSize.x + ", " + displayRealSize.y);
+            int mX = displayRealSize.x/2;
+            //int mY = displayRealSize.y - oppoLocation + mW;
+            int mY = displayRealSize.y - oppoLocation;
+
+            Slog.d("PHH-Enroll", "\tfacola at  " + mX + ", " + mY);
+            udfpsProps = new int[3];
+            udfpsProps[0] = (int)mX;
+            udfpsProps[1] = (int)mY;
+            udfpsProps[2] = (int)mW;
+            isUdfps = true;
+        }
+
+        // Asus ZF8
+        if(android.os.SystemProperties.get("ro.vendor.build.fingerprint").contains("ASUS_I006D")) {
+            udfpsProps = new int[3];
+            udfpsProps[0] = displayRealSize.x/2;;
+            udfpsProps[1] = 1741;
+            udfpsProps[2] = 110;
+            isUdfps = true;
+        }
+
+        // ROG Phone 3
+        if(android.os.SystemProperties.get("ro.vendor.build.fingerprint").contains("ASUS_I003")) {
+            udfpsProps = new int[3];
+            udfpsProps[0] = displayRealSize.x/2;;
+            udfpsProps[1] = 1752;
+            udfpsProps[2] = 110;
+            isUdfps = true;
+        }
+
+        // Redmagic 5g
+        if("NX659J".equals(android.os.SystemProperties.get("ro.product.vendor.model"))) {
+            udfpsProps = new int[3];
+            udfpsProps[0] = displayRealSize.x/2;;
+            udfpsProps[1] = 1984;
+            udfpsProps[2] = 95;
+            isUdfps = true;
+        }
+
+	if(udfpsProps.length > 0) {
+	    Slog.d("PHH-Enroll", "Samsung got udfps infos " + udfpsProps[0] + ", " + udfpsProps[1] + ", " + udfpsProps[2]);
+	}
+
         final @FingerprintSensorProperties.SensorType int sensorType;
         if (isUdfps) {
             sensorType = FingerprintSensorProperties.TYPE_UDFPS_OPTICAL;
diff --git a/services/core/java/com/android/server/biometrics/sensors/fingerprint/UdfpsHelper.java b/services/core/java/com/android/server/biometrics/sensors/fingerprint/UdfpsHelper.java
index 29661d46f328..e97f70d7bffd 100644
--- a/services/core/java/com/android/server/biometrics/sensors/fingerprint/UdfpsHelper.java
+++ b/services/core/java/com/android/server/biometrics/sensors/fingerprint/UdfpsHelper.java
@@ -23,6 +23,10 @@ import android.hardware.fingerprint.FingerprintManager;
 import android.os.RemoteException;
 import android.util.Slog;
 
+import vendor.goodix.hardware.biometrics.fingerprint.V2_1.IGoodixFingerprintDaemon;
+
+import java.io.PrintWriter;
+
 /**
  * Contains helper methods for under-display fingerprint HIDL.
  */
@@ -30,37 +34,120 @@ public class UdfpsHelper {
 
     private static final String TAG = "UdfpsHelper";
 
+    private static void writeFile(String path, String value) {
+        try {
+            PrintWriter writer = new PrintWriter(path, "UTF-8");
+            writer.println(value);
+            writer.close();
+        } catch(Exception e) {
+            android.util.Log.d("PHH", "Failed writing to " + path + ": " + value);
+        }
+    }
+
+
     public static void onFingerDown(IBiometricsFingerprint daemon, int x, int y, float minor,
             float major) {
         android.hardware.biometrics.fingerprint.V2_3.IBiometricsFingerprint extension =
-                android.hardware.biometrics.fingerprint.V2_3.IBiometricsFingerprint.castFrom(
-                        daemon);
-        if (extension == null) {
-            Slog.v(TAG, "onFingerDown | failed to cast the HIDL to V2_3");
+            android.hardware.biometrics.fingerprint.V2_3.IBiometricsFingerprint.castFrom(
+                    daemon);
+        if (extension != null) {
+            try {
+                extension.onFingerDown(x, y, minor, major);
+                return;
+            } catch (RemoteException e) {
+                Slog.e(TAG, "onFingerDown | RemoteException: ", e);
+            }
+        }
+
+        try {
+            // Asus goodix commands
+            IGoodixFingerprintDaemon goodixDaemon = IGoodixFingerprintDaemon.getService();
+            if(android.os.SystemProperties.get("ro.vendor.build.fingerprint").contains("ASUS")) {
+                goodixDaemon.sendCommand(200001, new java.util.ArrayList<Byte>(), (returnCode, resultData) -> {
+                    Slog.e(TAG, "Goodix send command returned code "+ returnCode);
+                });
+            } else {
+                //UI READY
+                goodixDaemon.sendCommand(0x600, new java.util.ArrayList<Byte>(), (returnCode, resultData) -> {
+                    Slog.e(TAG, "Goodix send command returned code "+ returnCode);
+                });
+                goodixDaemon.sendCommand(1607, new java.util.ArrayList<Byte>(), (returnCode, resultData) -> {
+                    Slog.e(TAG, "Goodix send command returned code "+ returnCode);
+                });
+            }
             return;
+        } catch(Throwable t) {
+            Slog.e(TAG, "Tried sending goodix daemon cmd failed", t);
         }
 
         try {
-            extension.onFingerDown(x, y, minor, major);
-        } catch (RemoteException e) {
-            Slog.e(TAG, "onFingerDown | RemoteException: ", e);
+            vendor.oplus.hardware.biometrics.fingerprint.V2_1.IBiometricsFingerprint fp = vendor.oplus.hardware.biometrics.fingerprint.V2_1.IBiometricsFingerprint.getService();
+            writeFile("/sys/kernel/oppo_display/notify_fppress", "1");
+            writeFile("/sys/kernel/oplus_display/oplus_notify_fppress", "1");
+            fp.touchDown();
+        } catch(Throwable t) {
+            Slog.e(TAG, "Tried sending oplus daemon cmd failed", t);
         }
+
+        try {
+            vendor.oppo.hardware.biometrics.fingerprint.V2_1.IBiometricsFingerprint fp = vendor.oppo.hardware.biometrics.fingerprint.V2_1.IBiometricsFingerprint.getService();
+            writeFile("/sys/kernel/oppo_display/notify_fppress", "1");
+            writeFile("/sys/kernel/oplus_display/oplus_notify_fppress", "1");
+            fp.touchDown();
+        } catch(Throwable t) {
+            Slog.e(TAG, "Tried sending oplus daemon cmd failed", t);
+        }
+
+        Slog.v(TAG, "onFingerDown | failed to cast the HIDL to V2_3");
     }
 
     public static void onFingerUp(IBiometricsFingerprint daemon) {
         android.hardware.biometrics.fingerprint.V2_3.IBiometricsFingerprint extension =
-                android.hardware.biometrics.fingerprint.V2_3.IBiometricsFingerprint.castFrom(
-                        daemon);
-        if (extension == null) {
-            Slog.v(TAG, "onFingerUp | failed to cast the HIDL to V2_3");
+            android.hardware.biometrics.fingerprint.V2_3.IBiometricsFingerprint.castFrom(
+                    daemon);
+        if (extension != null) {
+            try {
+                extension.onFingerUp();
+                return;
+            } catch (RemoteException e) {
+                Slog.e(TAG, "onFingerUp | RemoteException: ", e);
+            }
+        }
+
+        try {
+            IGoodixFingerprintDaemon goodixDaemon = IGoodixFingerprintDaemon.getService();
+            if(android.os.SystemProperties.get("ro.vendor.build.fingerprint").contains("ASUS")) {
+                goodixDaemon.sendCommand(200003, new java.util.ArrayList<Byte>(), (returnCode, resultData) -> {
+                    Slog.e(TAG, "Goodix send command returned code " + returnCode);
+                });
+            } else {
+                goodixDaemon.sendCommand(0x601, new java.util.ArrayList<Byte>(), (returnCode, resultData) -> {
+                    Slog.e(TAG, "Goodix send command returned code "+ returnCode);
+                });
+            }
             return;
+        } catch(Throwable t) {
+            Slog.e(TAG, "Tried sending goodix daemon cmd failed", t);
+        }
+
+        try {
+            vendor.oplus.hardware.biometrics.fingerprint.V2_1.IBiometricsFingerprint fp = vendor.oplus.hardware.biometrics.fingerprint.V2_1.IBiometricsFingerprint.getService();
+            writeFile("/sys/kernel/oppo_display/notify_fppress", "0");
+            writeFile("/sys/kernel/oplus_display/oplus_notify_fppress", "0");
+            fp.touchUp();
+        } catch(Throwable t) {
+            Slog.e(TAG, "Tried sending oplus daemon cmd failed", t);
         }
 
         try {
-            extension.onFingerUp();
-        } catch (RemoteException e) {
-            Slog.e(TAG, "onFingerUp | RemoteException: ", e);
+            vendor.oppo.hardware.biometrics.fingerprint.V2_1.IBiometricsFingerprint fp = vendor.oppo.hardware.biometrics.fingerprint.V2_1.IBiometricsFingerprint.getService();
+            writeFile("/sys/kernel/oppo_display/notify_fppress", "0");
+            writeFile("/sys/kernel/oplus_display/oplus_notify_fppress", "0");
+            fp.touchUp();
+        } catch(Throwable t) {
+            Slog.e(TAG, "Tried sending oplus daemon cmd failed", t);
         }
+        Slog.v(TAG, "onFingerUp | failed to cast the HIDL to V2_3");
     }
 
     public static boolean isValidAcquisitionMessage(@NonNull Context context,
-- 
2.25.1

